File: 23_merge_k_sorted_lists/main.py
# -*- coding: utf-8 -*-


class ListNode(object):

    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):

    def merge_k_sorted_lists(self, lists):
        size = len(lists)
        if size == 0:
            return None
        while size > 1:
            for i in range(size / 2):
                lists[i] = self.merge_two_sorted_lists(lists[i], lists[size - i - 1])
            size = int((size + 1) / 2)
        return lists[0]

    def merge_two_sorted_lists(self, l1, l2):
        if l1 is None:
            return l2
        if l2 is None:
            return l1
        if l1.val < l2.val:
            l1.next = self.merge_two_sorted_lists(l1.next, l2)
            return l1
        else:
            l2.next = self.merge_two_sorted_lists(l1, l2.next)
            return l2

File: 23_merge_k_sorted_lists/main.cpp
#include <iostream>
#include <vector>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* mergeKSortedLists(std::vector<ListNode*>& lists) {
        int len = lists.size();
        if (len == 0) return NULL;

        while (len > 1) {
            for (int i = 0; i < len / 2; ++i) lists[i] = mergeTwoSortedLists(lists[i], lists[len - 1 - i]);
            len = (len + 1) / 2;
        }
        return lists[0];
    }

    ListNode* mergeTwoSortedLists(ListNode* l1, ListNode* l2) {
        if (l1 == NULL) return l2;
        if (l2 == NULL) return l1;
        if (l1->val < l2->val) {
            l1->next = mergeTwoSortedLists(l1->next, l2);
            return l1;
        } else {
            l2->next = mergeTwoSortedLists(l1, l2->next);
            return l2;
        }
    }
};

int main() {
    std::cout << "23" << std::endl;
}
